python/ray/data/datasource/file_based_datasource.py

```diff
@@ -909,3 +927,44 @@ def _fetch_metadata_parallel(
         fetch_tasks.append(remote_fetch_func.remote(uri_chunk))
     results = metadata_fetch_bar.fetch_until_complete(fetch_tasks)
     yield from itertools.chain.from_iterable(results)
+
+
+def _open_file_with_retry(
+    file_path: str,
+    open_file: Callable[[], "pyarrow.NativeFile"],
+) -> "pyarrow.NativeFile":
+    """Open file with an exponential backoff retry strategy.
+
+    This is to avoid transient task failure with remote storage (such as S3),
+    when the remote storage throttles the requests.
+    """
+    import random
+    import time
+
+    if OPEN_FILE_MAX_ATTEMPTS < 1:
+        raise ValueError(
+            "OPEN_FILE_MAX_ATTEMPTS cannot be negative or 0. Get: "
+            f"{OPEN_FILE_MAX_ATTEMPTS}"
+        )
+
+    for i in range(OPEN_FILE_MAX_ATTEMPTS):
+        try:
+            return open_file()
+        except Exception as e:
+            error_message = str(e)
+            is_retryable = any(
+                [error in error_message for error in OPEN_FILE_RETRY_ON_ERRORS]
+            )
+            if is_retryable and i + 1 < OPEN_FILE_MAX_ATTEMPTS:
+                # Retry with binary expoential backoff with random jitter.
+                backoff = min(
+                    (2 ** (i + 1)) * random.random(),
```

Comment:
This is same as AWS retry behavior - https://docs.aws.amazon.com/sdkref/latest/guide/feature-retry-behavior.html .

<img width="867" alt="Screen Shot 2023-08-23 at 4 53 37 PM" src="https://github.com/ray-project/ray/assets/4629931/e49e810d-bdd8-4a73-a416-89218a59a529">